Update the contribution guide (#76)
kgateway-dev/community: CONTRIBUTING.md

Here are some of the ways you can contribute:

- [Requirements for PRs](#requirements-for-prs)
  - [DCO](#dco)
  - [Commit Signing](#commit-signing)
  - [Code review guidelines](#code-review-guidelines)
- [Ways to contribute](#ways-to-contribute)
  - [Report Security Vulnerabilities](#report-security-vulnerabilities)
  - [File issues](#file-issues)
  - [Find something to work on](#find-something-to-work-on)
  - [Small changes (bug fixes)](#small-changes-bug-fixes)
  - [Large changes (features, refactors)](#large-changes-features-refactors)
- [Documentation](#documentation)
- [Get in touch](#get-in-touch)

To understand contributor roles, refer to the [contributor ladder guide](CONTRIBUTOR_LADDER.md).

## Requirements for PRs

> Contributing to open source can be a daunting task, especially if you are a new contributor and are not yet familiar with the workflows commonly used by open source projects. Kgateway uses a "fork and pull request" approach. This means that contributors will typically create their own personal fork of a code repository in GitHub and push their contributions to a branch in their own fork first. Once they are ready to contribute they can open a pull request (PR) against the project's repository which will be reviewed by the project. Reviews will typically include iterations of suggestions and changes. This is totally normal, don't be discouraged if asked to make changes to your contribution.

Reviewer: ```suggestion
Contributing to open source can be a daunting task, especially if you are a new contributor and are not yet familiar with the workflows commonly used by open source projects. Kgateway uses a "fork and pull request" approach. This means that contributors will typically create their own personal fork of a code repository in GitHub and push their contributions to a branch in their own fork first. Once they are ready to contribute they can open a pull request (PR) against the project's repository which will be reviewed by the project maintainers. Reviews will typically include iterations of suggestions and changes. This is totally normal, don't be discouraged if asked to make changes to your contribution.
```

> Contributing to open source can be a daunting task, especially if you are a new contributor and are not yet familiar with the workflows commonly used by open source projects. Kgateway uses a "fork and pull request" approach. This means that contributors will typically create their own personal fork of a code repository in GitHub and push their contributions to a branch in their own fork first. Once they are ready to contribute they can open a pull request (PR) against the project's repository which will be reviewed by the project. Reviews will typically include iterations of suggestions and changes. This is totally normal, don't be discouraged if asked to make changes to your contribution.

Reviewer: also, it won't let me add a comment there, but in the paragraph below this, there's a `#big-prs` link that doesn't go anywhere

> The easiest way to signoff on your contributions is to use the `-s` flag on your git commit command when you check in code. `git commit -s -m "description of my excellent contribution"`. If your forget to signoff for a commit our automation will flag your PR by failing the DCO check. If this happens you will need to go back and signoff on previous commits. `git rebase --signoff main` can be used to add signoffs to previous commits. In our example `main` is used, which means this command will rewrite the git history of your current branch adding signoffs to commits visible from `main` (not inclusive). Please be aware that rewriting commit history does carry some risk and if the commits you are rewriting we already pushed to a remote you will need to force push the rewritten history.

Reviewer: i think we got rid of the DCO requirement, do we want to remove this section, or just say it's recommended but not required?

Reviewer 2: CNCF projects require either DCO or CLA

Author: I left this in for now. Although it does mention that the PRs have DCO checks, which I'm not sure is accurate?

Reviewer: we removed the DCO checks from PRs because it was too much of a hassle :(

Reviewer: there's a tracking issue for it https://github.com/kgateway-dev/community/issues/30

Reviewer 3: Yup, there's disagreement over whether one-and-done CLA or DCO on every single commit now and forever is worse for contributors. IMO if you are only ever going to make a single commit to any kgw repo then DCO is better but if you ever make a second commit then you're playing a losing hand vs a one-and-done CLA process. I'm not sure it really makes sense to prioritize flyby contribs by creating permanent low-grade pain for everyone else but it seems OSS community wide I am in the minority on that one so if we get stuck with DCO I'll live with it (or quit working on kgw due to the hassle and work on Istio instead 😜)

Author: Thanks for the context. I updated the wording so it does not explicitly state something incorrect right now (that we check PRs right now) and added either DCO or CLA.

### Commit signing

In order to help ensure the integrity of our code the kgateway project requires that your commits are signed and verified. For more details as well as instructions for setting up signing please review the [github docuementation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) about signing and verification of commits.

### Code review guidelines

All code must be reviewed by at least one [maintainer](https://github.com/kgateway-dev/community/blob/main/MAINTAINERS.md). Key requirements:

1. **Code Style**
   
   - Follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments).
   - Follow [Effective Go](https://golang.org/doc/effective_go).
   - Run `make analyze` to check for common issues before submitting.

2. **Testing**

   - Add unit tests for new functionality.
   - Ensure existing tests pass.
   - Include integration and e2e tests when needed.

3. **Documentation**
   
   - Update relevant documentation.
   - Include code comments for non-obvious logic.
   - Update API documentation if changing interfaces.

4. **Change management**
   

>    - Do not make breaking changes.

Reviewer 2: unrealistic?  "Minimize the number of breaking changes", "As a rule, avoid breaking changes" or such?

Reviewer: yeah i think we can remove this line. the line below addresses what to do if you need to make a breaking change

   - If your PR potentially introduces a breaking change, make sure to call it out in the PR description and that it follows the agreed-upon design from your issue.
   - Please address or acknowledge review comments on your PRs.
   - If you are reviewing, please try to leave helpful, polite, substantive comments. Minor preferences can be called out with `nit`.
   - Write a PR description that can be pulled into a changelog.

## Ways to contribute

Thanks for your interest in contributing to kgateway! We have a few different ways you can get involved.

### Report Security Vulnerabilities

> If you would like to report a security issue, please refer to the [kgateway documentation site](https://k8sgateway.io/docs/reference/vulnerabilities/). 

Reviewer: ```suggestion
If you would like to report a security issue, please refer to the [kgateway documentation site](https://kgateway.dev/docs/reference/vulnerabilities/). 
```

### File issues

> To file a bug or feature request in the [kgateway](https://github.com/kgateway-dev/kgateway) GitHub repo:

Reviewer 2: ```suggestion
To file a bug or feature request in the [kgateway GitHub repo](https://github.com/kgateway-dev/kgateway):
```

1. Search existing issues first.
2. If no existing issue addresses your case, create a new one.
3. Use issue templates when available.
4. Add information or react to existing issues, such as a thumbs-up 👍 to indicate agreement.

### Find something to work on

> The project uses [GitHub issues](https://github.com/kgateway-dev/kgateway/issues) to track bugs and features. Issues labeled with the `good first issue` label are a great place to start.

Reviewer: ```suggestion
The project uses [GitHub issues](https://github.com/kgateway-dev/kgateway/issues) to track bugs and features. Issues labeled with the `Good First Issue` label are a great place to start.
```
nit: in the kgateway repo we currently have the label capitalized

Author: How proper of us; thanks for the catch 😄

Additionally, the project has a [milestone](https://github.com/kgateway-dev/kgateway/milestones) for the next release. Any issues labeled with a milestone are a great source of things to work on. If an issue has not been assigned to a milestone, you can ask to work on it by leaving a comment on the issue.

Flaky tests are a common source of issues and a good place to start contributing to the project. You can find these issues by filtering with the `Type: CI Test Flake` label. If you see a test that is failing regularly, you can leave a comment asking if someone is working on it.

### Small changes (bug fixes)

For small changes (less than 100 lines of code):

1. Open a pull request.
2. Ensure tests verify the fix.
3. If needed, [update the documentation](#documentation).

### Large changes (features, refactors)

Large features often touch many files, extend many lines of code, and often cover issues such as:

* Large bug fixes
* New features
* Refactors of the existing codebase

For large changes:
1. **Open an issue first**: Open an issue about your bug in the [kgateway](https://github.com/kgateway-dev/kgateway) repo.

> 2. **Message us on Slack**: Reach out to us to discuss your proposed changes in our [CNCF slack channel, #kgateway](https://cloud-native.slack.com/archives/C080D3PJMS4).

Reviewer 2: ```suggestion
2. **Message us on Slack**: Reach out to us to discuss your proposed changes in our [CNCF Slack channel, #kgateway](https://cloud-native.slack.com/archives/C080D3PJMS4).
```

3. **Agree on implementation plan**: Write a plan for how this feature or bug fix should be implemented. Should this be one pull request or multiple incremental improvements? Who is going to do each part? Discuss it with us on Slack or join our [community meeting](https://calendar.google.com/calendar/u/1?cid=ZDI0MzgzOWExMGYwMzAxZjVkYjQ0YTU0NmQ1MDJmODA5YTBjZDcwZGI4ZTBhZGNhMzIwYWRlZjJkOTQ4MzU5Y0Bncm91cC5jYWxlbmRhci5nb29nbGUuY29t)
4. **Submit a draft PR**: It's important to get feedback as early as possible to ensure that any big improvements end up being merged. Open a draft pull request from your fork, label it `work in progress`, and start getting feedback.
5. **Review**: At least one maintainer should sign off on the change before it’s merged. Look at the following [Code review](#code-review-guidelines) section to learn about what we're looking for.
6. **Close out**: A maintainer will merge the PR and let you know about the next release plan.

## Documentation

> The kgateway documentation lives in a separate repository: [kgateway-dev/kgateway.dev](https://github.com/kgateway-dev/kgateway.dev). For more details, see the [Docs contribution guide](https://kgateway.dev/docs/reference/contribution/).

Reviewer 2: ```suggestion
The kgateway documentation [lives in a separate repository: `kgateway-dev/kgateway.dev`](https://github.com/kgateway-dev/kgateway.dev). For more details, see the [Docs contribution guide](https://kgateway.dev/docs/reference/contribution/).
```
